106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,507 @@
+Hence, these are not goals of this proposal and could be addressed later in separate proposals.
+Following the above approach will provide several advantages:
+* we ensure that the operator controls all rebalance and cluster remediation operations.
+* using the existing `KafkaRebalance` CR system gives more visibility into what is happening and when, which (as we don't support the Cruise Control UI) enhances observability and will also aid in debugging.
+
+### `imbalance` mode in Strimzi's auto-rebalancing feature
+
+The [`auto-rebalancing`](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) feature in Strimzi allows the operator to run a rebalance automatically when a Kafka cluster is scaled up (by adding brokers) or scaled down (by removing brokers).
+
+Auto-rebalancing in Strimzi currently supports two modes:
+* `add-brokers` - auto-rebalancing on scale up
+* `remove-brokers` - auto-rebalancing on scale down
+
+To leverage the automated rebalance on imbalanced cluster (those with detected Goal violations), we will be introducing a new mode to the auto-rebalancing feature.
+The new mode will be called `imbalance`, which means that cluster imbalance was detected and rebalancing should be applied to all the brokers.
+The imbalance mode is configured by creating a new object in the `spec.cruiseControl.autoRebalance` list with its `mode` field set to `imbalance` and the corresponding rebalancing configuration is defined as a reference to a template `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+The `template` field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration (i.e. the same used for unmodified manual `KafkaRebalance` invocations).
+To provide users more flexibility, they only have to configure the auto-rebalance modes they wish to use whether it be `add-brokers`, `remove-brokers`, or `imbalance`.
+When the auto-rebalance configuration is set with `imbalance` mode enabled, the operator will trigger a partition rebalance whenever a goal violation is detected by the anomaly detector.
+To trigger the auto-rebalance, the operator must know that the cluster is imbalanced due to some goal violation anomaly.
+We will create our own custom notifier named `StrimziCruiseControlNotifier` to do this.
+This notifier's job will be to update the operator regarding the goal violations so that the operator can trigger a rebalance (see section [StrimziCruiseControlNotifier](./106-auto-rebalance-on-imbalanced-clusters.md#strimzicruisecontrolnotifier)).
+With this proposal, we are only going to support auto-rebalance on imbalanced cluster.
+We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approach.
```

Comment:
```suggestion
We also plan to implement the same for topic and metrics related issues, but it will be part of future work since their implementation require different approaches.
```